homepage/content/blog/2021-03-05-switchMap-concatMap-mergemap-exhaustmap-explained.md

```diff
@@ -0,0 +1,264 @@
+
+```ts
+import { Observable } from "rxjs";
+
+anyLongRunningOp(value: string) {
+  return new Observable(observer => {
+    setTimeout(() => {
+      observer.next(value);
+      observer.complete();
+    }, 2000);
+  });
+}
+```
+
+This method does nothing but wait for two seconds and then return a value. Could be an HTTP call in the real world or something else asynchronous which takes a little bit of time.
+
+Using this method would be like
+
+```ts
+// ...
+
+anyLongRunningOp('value').subscribe((result) => {
+  // after two seconds we get our subscribe called
+  console.log(result); // <<<< logs 'value'
+});
+```
+
+To explain `switchMap`, `mergeMap`, `concatMap` and `exhaustMap` let us _always_ assume that we call the method `anyLongRunningOp` _multiple times_! This is important. So we call it and immediately call the method again but the first call has not completed (hence the 2 seconds delay).
+
+To be able to emit multiple observables we can create a `Subject` where we can write values in with `next(...)` and process it with the operator we want to take a look at.
+
+```ts
+sub = new Subject<string>();
+
+fireEvents() {
+
+  // Here we react ot everything which is fired in the subject
+  this.sub
+    // Here we can take the operator we want to take a look at which returns the
+    // result from the anyLongRunningOp method which is the value itself
+    // (for the sake of simplicity). The `tap` only prints out what
+    // was just sent, does nothing to the stream!
+    .pipe(
+      tap(value => console.log("--> sent out", value)),
+      [OPERATOR_HERE]((value) => this.anyLongRunningOp(value)))
+    // We console.log the output, which is 'first' or 'second' or
+    .subscribe(value => console.log("<-- received", value));
+
+  // After subscribing we emit the two value in the observable, could also be more than that
+  this.sub.next('first');
+  this.sub.next('second');
+
+  console.log("-------");
+}
+
+anyLongRunningOp(value: string) {
+  return new Observable(observer => {
+    setTimeout(() => {
+      observer.next(value);
+      observer.complete();
+    }, 2000);
+  });
+}
+```
+
+The `fireEvents()` method emits in the `Subject` `sub` two times one value each. First the value `first` and immediately after the value `second`. This value then gets processed by the operator we want to look how it behaves when multiple values come in.
+
+## Similarities
+
+All of those operators can be piped to an observable and return a new observable. They keep the stream. They resolve the value but return a new observable. This makes them higher order mapping operators. You can subscribe to the outcome but initially `switchMap`, `mergeMap`, `concatMap` and `exhaustMap` return an observable which then can be processed further.

```

Comment:
You can subscribe to the outcome but initially `switchMap`, `mergeMap`, `concatMap` and `exhaustMap` return an observable which then can be processed further.

Vielleicht verstehe ich nicht ganz was du sagen willst, aber dieser satz gilt prinzipiell für jeden operator

Reply:
Wäre `return a _new_ observable` passender? Andere Operatoren wie `filter` oder `tap` geben kein neues Observable zurück, oder?

Reply:
doch tun sie :D Jeder operator erzeugt ein neues Observable, und managed dann die subscription vom source observable.

https://github.com/ReactiveX/rxjs/blob/6.5.5/src/internal/operators/map.ts#L50

in Observable.lift wird ein neues Observable erzeugt und zurück gegeben

Reply:
TIL! Danke!